sdk/servicebus/Microsoft.Azure.ServiceBus/src/Primitives/ConcurrentExpiringSet.cs

```diff
@@ -4,64 +4,71 @@
+        static readonly TimeSpan delayBetweenCleanups = TimeSpan.FromSeconds(30);
 
         public ConcurrentExpiringSet()
         {
             this.dictionary = new ConcurrentDictionary<TKey, DateTime>();
+            this.dictionaryAsCollection = dictionary;
+            _ = CollectExpiredEntriesAsync(tokenSource.Token);
         }
 
         public void AddOrUpdate(TKey key, DateTime expiration)
         {
             this.dictionary[key] = expiration;
-            this.ScheduleCleanup();
+            this.cleanupTaskCompletionSource.TrySetResult(true);
         }
 
         public bool Contains(TKey key)
         {
             return this.dictionary.TryGetValue(key, out var expiration) && expiration > DateTime.UtcNow;
         }
 
-        void ScheduleCleanup()
+        public void Clear()
         {
-            lock (this.cleanupSynObject)
-            {
-                if (this.cleanupScheduled || this.dictionary.Count <= 0)
-                {
-                    return;
-                }
-
-                this.cleanupScheduled = true;
-                Task.Run(async () => await this.CollectExpiredEntriesAsync().ConfigureAwait(false));
-            }
+            this.tokenSource.Cancel();
+            this.dictionary.Clear();
+            this.cleanupTaskCompletionSource.TrySetCanceled();
+            this.cleanupTaskCompletionSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+            this.tokenSource = new CancellationTokenSource();
+            _ = CollectExpiredEntriesAsync(tokenSource.Token);
         }
 
-        async Task CollectExpiredEntriesAsync()
+        async Task CollectExpiredEntriesAsync(CancellationToken token)
         {
-            await Task.Delay(delayBetweenCleanups);
-
-            lock (this.cleanupSynObject)
+            while (!token.IsCancellationRequested)
             {
-                this.cleanupScheduled = false;
-            }
+                try
+                {
+                    await cleanupTaskCompletionSource.Task.ConfigureAwait(false);
+                    this.cleanupTaskCompletionSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+                    await Task.Delay(delayBetweenCleanups, token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    return;
+                }
 
-            foreach (var key in this.dictionary.Keys)
-            {
-                if (DateTime.UtcNow > this.dictionary[key])
+                foreach (var kvp in this.dictionary)
                 {
-                    this.dictionary.TryRemove(key, out _);
+                    var expiration = kvp.Value;
+                    if (DateTime.UtcNow > expiration)
+                    {
+                        this.dictionaryAsCollection.Remove(kvp);
```

Comment:
I think this will modify the underlying collection while you are iterating over it which will throw an exception

Reply:
It is safe, see https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getenumerator